### ISSUE TITLE
Memory leak?

### ISSUE THREAD

**Issue by RMeli** [Amber-MD/pytraj#1495]
I'm not sure what's going on, but I noticed that if I try to extract a single frame from an out-of-memory trajectory (multiple times, for multiple trajectories) in order to save it on a separate file, RAM usage steadily increases. Since I'm working with a large data set, this causes the RAM to fill up and the whole application to crash.

This is a MWE that reproduces my problem:

```python
import pytraj as pt
import random

traj = pt.iterload("traj.nc", "top.parm7")

while True:
    r = int(random.uniform(0, 5000))
    print(r)
    protein = traj[[r], "!:LIG,WAT"]
    # Write protein to file
```

The trajectory contains 5000 frames and is around ~600MB, but the script above fills up the whole 64GB of RAM if run long enough. Any idea of what's going on? Am I doing something wrong?

**Comment by RMeli** [comment]
If I break the loop manually (with `ctrl-C`) the RAM is not freed. It's not freed if I set `traj` or `protein` to `None` or if I explicitly delete such variables (after manually breaking the loop). The RAM is only freed when I exit the Python interpreter (`ipython`).

**Comment by hainm** [comment]
It looks like a (another) bug. Thanks for reporting and I really appreciate it.

**Comment by RMeli** [comment]
No problem! Please let me know if I can be on any help solving the issue. It is a major blocker in the pipeline I'm working on at the moment and therefore I'm very keen to help fix it, if possible.

**Comment by hainm** [comment]
ok, I can make this my top priority.

**Comment by hainm** [comment]
@RMeli Does below code help you (using `iterframe`)

```
In [22]: for frame in traj.iterframe(frame_indices=[1, 2, 3]): 
    ...:     newt = pt.Trajectory(top=traj.top, xyz=np.array([frame.xyz])) 
    ...:     st = newt['!:WAT,LIG'] 
    ...:     print(st) 
```

**Comment by RMeli** [comment]
Thanks for the help @hainm! Your solution does not seem to solve the problem, unfortunately. The RAM is still increasing constantly when I use it as follows:

```
while True:
    r = int(random.uniform(0, 5000))
    for frame in traj.iterframe(frame_indices=[r,]):
        newt = pt.Trajectory(top=traj.top, xyz=np.array([frame.xyz]))
        st = newt['!:WAT,LIG']
```

In my pipeline, I need to load multiple different trajectories in sequence (corresponding to the external `while True` loop ) and extract from those a given number of frames which need to be saved in separate PDB files. Something like the following (pseudocode):

```
for trajname in trajnames:
    traj = pt.iterload(f"{trajname}.nc", f"{trajname}.parm7")

    idxs = FRAMES_TO_SAVE(traj)

    for idx in idxs:
        for frame in traj.iterframe(frame_indices=[idx,]):
            newt = pt.Trajectory(top=traj.top, xyz=np.array([frame.xyz]))
            st = newt['!:WAT,LIG']
            pt.write_traj(filename=f"{trajname}.{idx}.pdb", traj=st)
```

**Comment by hainm** [comment]
```
    for idx in idxs:
        for frame in traj.iterframe(frame_indices=[idx,]):
```

Why don't you put the `idxs` in the `iterframe` and try again?

```
for frame in traj.iterframe(frame_indices=idxs):
    bla bla
```

**Comment by RMeli** [comment]
The reason I'm not doing this is related to #1494; I need to be able to save a single frame in a PDB file. Therefore I have to do something like the following:

```
for frame in traj.iterframe(frame_indices=idxs):
    newt = pt.Trajectory(top=traj.top, xyz=np.array([frame.xyz]))
    st = newt['!:WAT,LIG']
    
    for idx in idxs:
        pt.write_traj(f"{trajname}.{idx}.pdb", traj=st, frame_indices=[idx,])
```

I'll try to see if `pt.write_traj` works properly in this particular case, where `st` is a new slice of trajectory and not the original one obtained `pt.iterload`.

**Comment by hainm** [comment]
Uhm, in my example `st` is a Trajectory with a single frame corresponding to your each index in `idxs`

```
for index, frame in enumerate(traj.iterframe(frame_indices=idxs)):
    newt = pt.Trajectory(top=traj.top, xyz=np.array([frame.xyz]))
    st = newt['!:WAT,LIG']
    idx = idxs[index]
    pt.write_traj(f{"trajname}.{idx}.pdb", traj=st)
```

**Comment by RMeli** [comment]
Yes, sorry! I was in a rush and did not think that through before replying. I'm now refactoring my pipeline to reflect your last reply.

**Comment by RMeli** [comment]
I just tested it with my refactored pipeline and with a simple MWE. Unfortunately, in both cases the RAM usage still increases steadily.

Below is the MWE, which is now very similar to the current pipeline:

```python
import pytraj as pt                                                                                                                                                                   
import numpy as np                                                                                                                                                                    

traj = pt.iterload("traj.nc", "top.parm7")                                                                                             

while True: 
     idxs = np.random.randint(0, 5000, size=100) 
     print(idxs) 
     for idx, frame in zip(idxs, traj.iterframe(frame_indices=idxs)): 
         newt = pt.Trajectory(top=traj.top, xyz=np.array([frame.xyz])) 
         protein = newt["!:LIG,WAT"] 
```

**Comment by hainm** [comment]
thanks @RMeli, I could (sandly) reproduce your issue with above code. I will try to investigate or find another (of another) workaround. cheers.

**Comment by RMeli** [comment]
Thanks @hainm , your help is really appreciated! I recently started digging into the code as well, I'll report here if I find any suspect for this problem.

**Comment by hainm** [comment]
Please see the workaround. Instead of using the `strip` or slicing stuff, just get the atom indices and slice the coordinates directly. :D

```python
import pytraj as pt                                                                                                                                                                   
import numpy as np                                                                                                                                                                    

traj = pt.iterload("sim.nc", "sim.prmtop")                                                                                             

stripped_top = traj.top.copy()
stripped_top.strip(':POPC')

atom_indices = pt.select_atoms(traj.top, '!:POPC')
print(atom_indices)

while True:
     idxs = np.random.randint(0, 39, size=10) 
     for idx, frame in zip(idxs, traj.iterframe(frame_indices=idxs)): 
         xyz = np.array([frame.xyz])[:, atom_indices]
         newt = pt.Trajectory(top=stripped_top, xyz=xyz)
         print(newt)
```

**Comment by hainm** [comment]
It turns out that it's `Topology.strip` leaks the memory.

```
import pytraj as pt                                                                                                                                                                   
import numpy as np                                                                                                                                                                    

ori_top = pt.load_topology("sim.prmtop")                                                                                             

i = 100000
while i > 0:
     top = ori_top.copy()
     top.strip(':POPC')
     i -= 1
```

[This code](https://github.com/Amber-MD/pytraj/issues/1495#issuecomment-557287219) works because the stripped topology is only created once (vs. traj.strip(...) or traj[...] will do the topology stripping).

**Comment by hainm** [comment]
In this source code, the topology pointer (to cpptraj's Topology object) points to different object but the pointed object is never be freed. :D 

https://github.com/Amber-MD/pytraj/blob/66e181535900dbac6befb357302a3bf8f6c6ceb3/pytraj/topology/topology.pyx#L595

**Comment by RMeli** [comment]
Amazing, thank you very much for the help! 

It is better to re-write my pipeline with the workaround, wait for a fix of the leak your spotted or provide a fix myself?

**Comment by hainm** [comment]
> It is better to re-write my pipeline with the workaround,

Please do this. Fixing the leak isn't easy IMO. Cheers.

**Comment by RMeli** [comment]
Ok, thank you very much again for your help!

I just tested the workaround on my pipeline and everything seems to run with low RAM usage. =)

**Comment by drroe** [comment]
@hainm Just to be clear is this a pytraj issue or something I need to tackle in cpptraj as well?

**Comment by hainm** [comment]
@drroe  it's a pytraj's issue (pytraj does not properly free memory of cpptraj's Topology object). Thanks Dan.

**Comment by hainm** [comment]
Thanks @RMeli  for your detail report. The issue is fixed (hopefully) in #1539 
(https://github.com/Amber-MD/pytraj/pull/1539/files#diff-a763fc00fcaac2bc07716b5027d5dbed)

It's clearly due to my inexperience with C++'s pointer. :D

**Comment by RMeli** [comment]
Thank you for the fix @hainm !
